lib/jellyfish/component/hls/request_handler.ex

```diff
@@ -0,0 +1,208 @@
+  end
+
+  def stop(room_id) do
+    GenServer.cast(registry_id(room_id), :shutdown)
+  end
+
+  @impl true
+  def init(%{room_id: room_id, room_pid: room_pid}) do
+    Process.monitor(room_pid)
+    {:ok, %__MODULE__{room_id: room_id, room_pid: room_pid}}
+  end
+
+  @impl true
+  def handle_cast({:update_recent_partial, last_partial, manifest}, state) do
+    status = Map.fetch!(state, manifest)
+    state = Map.put(state, :manifest, update_status(status, last_partial))
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_cast({:is_partial_ready, partial, from, manifest}, state) do
+    state =
+      state
+      |> Map.fetch!(manifest)
+      |> handle_is_partial_ready(partial, from)
+      |> then(&Map.put(state, manifest, &1))
+
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_cast(:shutdown, state) do
+    {:stop, :normal, state}
+  end
+
+  @impl true
+  def terminate(_reason, %{room_id: room_id}) do
+    EtsHelper.remove_room(room_id)
+  end
+
+  @impl true
+  def handle_info({:DOWN, _ref, :process, pid, _reason}, %{room_pid: pid} = state) do
+    {:stop, :normal, state}
+  end
+
+  ###
+  ### PRIVATE FUNCTIONS
+  ###
+
+  defp wait_for_manifest_ready(room_id, partial, manifest) do
+    GenServer.cast(registry_id(room_id), {:is_partial_ready, partial, self(), manifest})
+
+    receive do
+      :manifest_ready ->
+        :ok
+    end
+  end
+
+  defp update_status(status, last_partial) do
+    {waiting_pids, status} =
+      status
+      |> Map.put(:last_partial, last_partial)
+      |> pop_in([:waiting_pids, last_partial])
+
+    send_partial_ready(waiting_pids)
+    status
+  end
+
+  defp handle_is_partial_ready(status, partial, from) do
+    if is_partial_ready(partial, status.last_partial) do
+      send(from, :manifest_ready)
+      status
+    else
+      waiting_pids =
+        Map.update(status.waiting_pids, partial, [from], fn pids_list ->
+          [from | pids_list]
+        end)
+
+      %{status | waiting_pids: waiting_pids}
+    end
+  end
+
+  defp registry_id(room_id), do: {:via, Registry, {Jellyfish.RequestHandlerRegistry, room_id}}
+
+  defp send_partial_ready(nil), do: nil
+
+  defp send_partial_ready(waiting_pids) do
+    Enum.each(waiting_pids, fn pid -> send(pid, :manifest_ready) end)
+  end
+
+  defp is_partial_ready(_partial, nil) do
+    false
+  end
+
+  defp is_partial_ready(partial, last_partial),
+    do:
+      is_last_segment_sn_greater(partial, last_partial) or
+        is_same_segment_greater_or_equal_partial_sn(partial, last_partial)
+
+  defp is_last_segment_sn_greater({segment_sn, _partial_sn}, {last_segment_sn, _last_partial_sn}),
+    do: last_segment_sn > segment_sn
+
+  defp is_same_segment_greater_or_equal_partial_sn(
+         {segment_sn, partial_sn},
+         {last_segment_sn, last_partial_sn}
+       ),
+       do: last_segment_sn == segment_sn and last_partial_sn >= partial_sn
```

Comment:
How about something like this?
```suggestion
  defp is_partial_ready({segment_sn, partial_sn}, {last_segment_sn, last_partial_sn}) do
    cond do
      last_segment_sn > segment_sn -> true
      last_segment_sn < segment_sn -> false
      true -> last_partial_sn >= partial_sn
    end
  end
```